[AutoDiff upstream] Add stdlib `Differentiable` conformances and derivatives.

dan-zheng opened the pull request:
Add `Differentiable` conformances and derivatives for stdlib types and functions to the `_Differentiation` module. Most upstreaming work done by @ematejska!

---

### `Differentiable` conformances
- `FloatingPoint`-conforming types: `Float16`, `Float`, `Double`, `Float80`
- `Array` where `Element: Differentiable`
- `SIMD`-conforming types where `Scalar: Differentiable & BinaryFloatingPoint, Scalar.TangentVector: BinaryFloatingPoint`

### Derivatives
- Floating-point operations: `+`, `+=`, `-`, `-=`, `*`, `*=`, `/`, `/=`
- Array operations: `Array.init(repeating:count:)`, `Array.subscript.get`, `Array.+(_:_:)`
- `SIMD` operations: `+`, `-`, `*`, `/`
- Most tgmath operations: `exp`, `log`, `cos`, `sin`, etc.

Differentiable collection higher-order functions:
- `Array.differentiableMap(_:)`, `Array.differentiableReduce(_:_:)`

Add end-to-end derivative tests.

---

Todos:

- `Differentiable.zeroTangentVector` protocol requirement and implementations.
  - TF-1238: to be upstreamed.
- `SIMD.sum` differentiation.
  - TF-1103: `@_alwaysEmitIntoClient` derivative registration.

ematejska committed:
First pass at upstreaming Differentiable conformances and derivatives

ematejska committed:
Merge branch 'master' of https://github.com/apple/swift

ematejska committed:
fixing up some tests and turning off SIMD extension which uses @_alwaysEmitIntoClient since that's not supported yet.

dan-zheng committed:
[AutoDiff upstream] Gardening.

- Standardize filenames: `XXXDifferentiation.swift`.
  - Use Pascal or snake case consistently.
- Formatting changes.

dan-zheng committed:
[AutoDiff upstream] Add `Float16` derivatives.

Add `@available` attributes for declarations using `Float16`.

dan-zheng committed:
Merge branch 'master' of github.com:apple/swift into autodiff_upstream_conformances_derivatives_master

dan-zheng committed:
[AutoDiff upstream] Deduplicate floating-point `Differentiable` conformances.

Move `Differentiable` conformances out of Differentiable.swift.
All `Differentiable` conformances exist in separate files.

dan-zheng committed:
[AutoDiff upstream] Fix stdlib differentiation tests.

Temporarily disable not-yet-supported differentiation tests:

- Forward-mode differentiation.
  - TF-1237: to be upstreamed.
- `Differentiable.zeroTangentVector`.
  - TF-1238: to be upstreamed.
- `SIMD.sum` differentiation.
  - TF-1103: `@_alwaysEmitIntoClient` derivative registration.

dan-zheng committed:
[AutoDiff upstream] Test differentiable collection higher-order functions.

Test `Array.differentiableMap(_:)` and `Array.differentiableReduce(_:_:)`.

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
Two tests fail with `-O` due to SR-12548: `SILCombiner::visitPartialApplyInst` SIL verification failure. I'll submit a fix soon.

ematejska commented on stdlib/public/Differentiation/FloatingPointDifferentiation.swift.gyb:
```diff
@@ -0,0 +1,275 @@
+//===----------------------------------------------------------------------===//
+
+import Swift
+import SwiftShims
```
We might not need to import SwiftShims on master. What do you think?  Aside from that looks good.

ematejska reviewed (commented):
Looks like there are a couple of test failures still.

dan-zheng committed:
Merge branch 'master' of github.com:apple/swift into autodiff-upstream-stdlib-differentiation

dan-zheng commented:
Test failures due to SR-12548 fixed in https://github.com/apple/swift/pull/30877.
@swift-ci Please test

dan-zheng committed:
Fix tests for 32-bit platforms.

- Remove unnecessary imports from test/AutoDiff/stdlib/simd.swift.
- Use platform-correct `Builtin` integer type in
  `PullbackEmitter::getArrayAdjointElementBuffer`.

dan-zheng commented:
Fixed test failures for 32-bit platforms.
@swift-ci Please test

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
@swift-ci Please test macOS

dan-zheng commented:
@swift-ci Please test Windows
